File: examples/node/ts/acknowledgementPayloads.ts
import * as readline from "readline/promises";
import * as fs from "fs";
import * as timer from "timers/promises";
import { RF24, PaLevel } from "@rf24/rf24";

const io = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

export class App {
  radio: RF24;
  counter: number;

  /**
   * @param radioNumber The number (0 or 1) that identifies the which radio is used in
   * this example that requires 2 radios.
   */
  constructor(radioNumber: number) {
    // The radio's CE Pin uses a GPIO number.
    // On Linux, consider the device path `/dev/gpiochip<N>`:
    //   - `<N>` is the gpio chip's identifying number.
    //     Using RPi4 (or earlier), this number is `0` (the default).
    //     Using the RPi5, this number is actually `4`.
    // The radio's CE pin must connected to a pin exposed on the specified chip.
    const cePin = 22; // for GPIO22
    // try detecting RPi5 first; fall back to default
    const gpioChip = fs.existsSync("/dev/gpiochip4") ? 4 : 0;

    // The radio's CSN Pin corresponds the SPI bus's CS pin (aka CE pin).
    // On Linux, consider the device path `/dev/spidev<a>.<b>`:
    //   - `<a>` is the SPI bus number (defaults to `0`)
    //   - `<b>` is the CSN pin (must be unique for each device on the same SPI bus)
    const csnPin = 0; // aka CE0 for SPI bus 0 (/dev/spidev0.0)

    // create a radio object for the specified hardware config:
    this.radio = new RF24(cePin, csnPin, {
      devGpioChip: gpioChip,
    });

    // initialize the nRF24L01 on the spi bus
    this.radio.begin();

    // For this example, we will use different addresses
    // An address needs to be a buffer object (bytearray)
    const address = [Buffer.from("1Node"), Buffer.from("2Node")];

    //set TX address of RX node into the TX pipe
    this.radio.openTxPipe(address[radioNumber]); // always uses pipe 0
    // set RX address of TX node into an RX pipe
    this.radio.openRxPipe(1, address[1 - radioNumber]); // using pipe 1

    // set the Power Amplifier level to -12 dBm since this test example is
    // usually run with nRF24L01 transceivers in close proximity of each other
    this.radio.setPaLevel(PaLevel.Low); // PaLevel.Max is default

    this.radio.allowAckPayloads(true);
    this.radio.setDynamicPayloads(true);

    this.counter = 0;
  }

  /**
   * The transmitting node's behavior.
   * @param count The number of payloads to send
   */
  async tx(count?: number) {
    this.radio.stopListening();
    // we'll use a DataView object to store our string and number into a bytearray buffer
    const outgoing = Buffer.from("Hello \0.");
    for (let i = 0; i < (count || 5); i++) {
      outgoing.writeUint8(this.counter, 7);
      const start = process.hrtime.bigint();
      const result = this.radio.send(outgoing);
      const end = process.hrtime.bigint();
      if (result) {
        const elapsed = (end - start) / BigInt(1000);
        process.stdout.write(
          `Transmission successful! Time to Transmit: ${elapsed} us. Sent: ` +
            `${outgoing.subarray(0, 6).toString()}${this.counter} `,
        );
        this.counter += 1;
        if (this.radio.available()) {
          const incoming = this.radio.read();
          const counter = incoming.readUint8(7);
          console.log(
            ` Received: ${incoming.subarray(0, 6).toString()}${counter}`,
          );
        } else {
          console.log("Received an empty ACK packet");
        }
      } else {
        console.log("Transmission failed or timed out!");
      }
      await timer.setTimeout(1000);
    }
  }

  /**
   * The receiving node's behavior.
   * @param duration The timeout duration (in seconds) to listen after receiving a payload.
   */
  rx(duration?: number) {
    this.radio.startListening();
    // we'll use a DataView object to store our string and number into a bytearray buffer
    const outgoing = Buffer.from("World \0.");
    outgoing.writeUint8(this.counter, 7);
    this.radio.writeAckPayload(1, outgoing);
    let timeout = Date.now() + (duration || 6) * 1000;
    while (Date.now() < timeout) {
      const hasRx = this.radio.availablePipe();
      if (hasRx.available) {
        const incoming = this.radio.read();
        const counter = incoming.readUint8(7);
        console.log(
          `Received ${incoming.length} bytes on pipe ${hasRx.pipe}: ` +
            `${incoming.subarray(0, 6).toString()}${counter} Sent: ` +
            `${outgoing.subarray(0, 6).toString()}${this.counter}`,
        );
        this.counter = counter;
        outgoing.writeUint8(counter + 1, 7);
        this.radio.writeAckPayload(1, outgoing);
        timeout = Date.now() + (duration || 6) * 1000;
      }
    }
    this.radio.stopListening(); // flushes TX FIFO when ACK payloads are enabled
  }

  /**
   * This function prompts the user and performs the specified role for the radio.
   */
  async setRole() {
    const prompt =
      "*** Enter 'T' to transmit\n" +
      "*** Enter 'R' to receive\n" +
      "*** Enter 'Q' to quit\n";
    io.resume();
    const input = (await io.question(prompt)).split(" ");
    io.pause();
    const role = input.shift() || "?";
    const params = Array<number>();
    for (let i = 0; i < input.length; ++i) {
      params.push(Number(input[i]));
    }
    switch (role.charAt(0).toLowerCase()) {
      case "t":
        await this.tx(...params);
        return true;
      case "r":
        this.rx(...params);
        return true;
      default:
        console.log(`'${role.charAt(0)}' is an unrecognized input`);
        return true;
      case "q":
        this.radio.powerDown();
        return false;
    }
  }
}

export async function main() {
  console.log(module.filename);

  io.resume();
  // to use different addresses on a pair of radios, we need a variable to
  // uniquely identify which address this radio will use to transmit
  // 0 uses address[0] to transmit, 1 uses address[1] to transmit
  const radioNumber = Number(
    (await io.question(
      "Which radio is this? Enter '1' or '0' (default is '0') ",
    )) == "1",
  );
  console.log(`radioNumber is ${radioNumber}`);
  io.pause();

  const app = new App(radioNumber);
  while (await app.setRole());
  io.close();
}

main();

File: examples/python/acknowledgement_payloads.py
"""
Simple example of using the library to transmit
and retrieve custom automatic acknowledgment payloads.

See documentation at https://nRF24.github.io/rf24-rs
"""

from pathlib import Path
import time
from rf24_py import RF24, PaLevel


class App:
    def __init__(self) -> None:
        # The radio's CE Pin uses a GPIO number.
        # On Linux, consider the device path `/dev/gpiochip<N>`:
        #   - `<N>` is the gpio chip's identifying number.
        #     Using RPi4 (or earlier), this number is `0` (the default).
        #     Using the RPi5, this number is actually `4`.
        # The radio's CE pin must connected to a pin exposed on the specified chip.
        ce_pin = 22  # for GPIO22
        # try detecting RPi5 first; fall back to default
        gpio_chip = 4 if Path("/dev/gpiochip4").exists() else 0

        # The radio's CSN Pin corresponds the SPI bus's CS pin (aka CE pin).
        # On Linux, consider the device path `/dev/spidev<a>.<b>`:
        #   - `<a>` is the SPI bus number (defaults to `0`)
        #   - `<b>` is the CSN pin (must be unique for each device on the same SPI bus)
        csn_pin = 0  # aka CE0 for SPI bus 0 (/dev/spidev0.0)

        # create a radio object for the specified hardware config:
        self.radio = RF24(ce_pin, csn_pin, dev_gpio_chip=gpio_chip)

        # For this example, we will use different addresses
        # An address need to be a buffer protocol object (bytearray)
        address = [b"1Node", b"2Node"]
        # It is very helpful to think of an address as a path instead of as
        # an identifying device destination

        # to use different addresses on a pair of radios, we need a variable to
        # uniquely identify which address this radio will use to transmit
        # 0 uses address[radio_number] to transmit, 1 uses address[not radio_number] to transmit
        radio_number = bool(
            int(input("Which radio is this? Enter '0' or '1'. Defaults to '0' ") or 0)
        )

        # initialize the nRF24L01 on the spi bus
        self.radio.begin()

        # set the Power Amplifier level to -12 dBm since this test example is
        # usually run with nRF24L01 transceivers in close proximity of each other
        self.radio.pa_level = PaLevel.Low  # PaLevel.Max is default

        # ACK payloads are dynamically sized, so we need to enable that feature also
        self.radio.dynamic_payloads = True

        # to enable the custom ACK payload feature
        self.radio.ack_payloads = True

        # set TX address of RX node into the TX pipe
        self.radio.open_tx_pipe(address[radio_number])  # always uses pipe 0

        # set RX address of TX node into an RX pipe
        self.radio.open_rx_pipe(1, address[not radio_number])  # using pipe 1

        self.counter = 0

        # for debugging
        # self.radio.print_details()

    def tx(self, count: int = 5):  # count = 5 will only transmit 5 packets
        """Transmits a payload every second and prints the ACK payload"""
        self.radio.as_tx()  # put radio in TX mode

        while count:
            # construct a payload to send
            buffer = b"Hello \x00" + bytes([self.counter])

            # send the payload and prompt
            start_timer = time.monotonic_ns()  # start timer
            result = self.radio.send(buffer)  # save the report
            end_timer = time.monotonic_ns()  # stop timer
            if result:
                # print timer results upon transmission success
                print(
                    "Transmission successful! Time to transmit:",
                    f"{int((end_timer - start_timer) / 1000)} us. Sent:",
                    f"{buffer[:6].decode('utf-8')}{self.counter}",
                    end=" ",
                )
                if self.radio.available():
                    # print the received ACK that was automatically sent
                    response = self.radio.read()
                    print(
                        f" Received: {response[:6].decode('utf-8')}{response[7:8][0]}"
                    )
                    self.counter += 1  # increment payload counter
                else:
                    print(" Received an empty ACK packet")
            else:
                print("Transmission failed or timed out")
            time.sleep(1)  # let the RX node prepare a new ACK payload
            count -= 1

    def rx(self, timeout: int = 6):
        """Prints the received value and sends an ACK payload"""
        self.radio.as_rx()  # put radio into RX mode, power it up

        # setup the first transmission's ACK payload
        buffer = b"World \x00" + bytes([self.counter])
        # we must set the ACK payload data and corresponding
        # pipe number [0,5]
        self.radio.write_ack_payload(1, buffer)  # load ACK for first response

        start = time.monotonic()  # start timer
        while (time.monotonic() - start) < timeout:
            has_payload, pipe_number = self.radio.available_pipe()
            if has_payload:
                received = self.radio.read()  # fetch 1 payload from RX FIFO
                print(
                    f"Received {len(received)} bytes on pipe {pipe_number}:",
                    f"{received[:6].decode('utf-8')}{received[7:8][0]} Sent:",
                    f"{buffer[:6].decode('utf-8')}{self.counter}",
                )
                start = time.monotonic()  # reset timer

                # increment counter from received payload
                self.counter = received[7:8][0] + 1
                # build a new ACK payload
                buffer = b"World \x00" + bytes([self.counter])
                self.radio.write_ack_payload(1, buffer)  # load ACK for next response

        # recommended behavior is to keep in TX mode while idle
        self.radio.as_tx()  # put radio in TX mode & flush unused ACK payloads

    def set_role(self):
        """Set the role using stdin stream. Timeout arg for slave() can be
        specified using a space delimiter (e.g. 'R 10' calls `slave(10)`)

        :return:
            - True when role is complete & app should continue running.
            - False when app should exit
        """
        user_input = (
            input(
                "*** Enter 'R' for receiver role.\n"
                "*** Enter 'T' for transmitter role.\n"
                "*** Enter 'Q' to quit example.\n"
            )
            or "?"
        )
        user_input = user_input.split()
        if user_input[0].upper().startswith("R"):
            self.rx(*[int(x) for x in user_input[1:2]])
            return True
        if user_input[0].upper().startswith("T"):
            self.tx(*[int(x) for x in user_input[1:2]])
            return True
        if user_input[0].upper().startswith("Q"):
            self.radio.power = False
            return False
        print(user_input[0], "is an unrecognized input. Please try again.")
        return True


if __name__ == "__main__":
    print(__file__)  # print example name

    app = App()
    try:
        while app.set_role():
            pass  # continue example until 'Q' is entered
    except KeyboardInterrupt:
        print(" Keyboard Interrupt detected. Exiting...")
        app.radio.power = False
